### PR TITLE
feat: identify pre onboarding v2 users

### PR DESCRIPTION
**Pull request by sshanzel**
To make it easier to identify if the user is valid for v2 or not, we calculate the dates on the API instead of FE to avoid any discrepancy.

### CODE REVIEW

File: src/routes/boot.ts [hunk 1]
```diff
@@ -359,9 +363,11 @@ const anonymousBoot = async (
     middleware ? middleware(con, req, res) : {},
     getAnonymousFirstVisit(req.trackingId),
   ]);
+  const isPreOnboardingV2 = new Date(firstVisit) < onboardingV2Requirement;
   return {
     user: {
       firstVisit,
+      isPreOnboardingV2,
```
Comment (rebelchris) [inline src/routes/boot.ts:370]: Just a suggestion why dont you immediately change the flag value in api?
We have done this before so from the frontend you'll only have to rely on flags

Comment (sshanzel) [reply src/routes/boot.ts:370]: Oh, never knew we've done that in the past, but we can definitely do that to make it easier.

File: src/routes/boot.ts [hunk 2]
```diff
@@ -359,14 +366,26 @@ const anonymousBoot = async (
...
   ]);
+  const isPreOnboardingV2 = firstVisit
+    ? new Date(firstVisit) < onboardingV2Requirement
+    : false;
+
+  const anonymousFlags = adjustAnonymousFlags(flags, [
+    {
+      replacement: 'control',
+      feature: 'onboarding_v2',
+      onCheckValidity: (value) => value !== 'control' && isPreOnboardingV2,
```
Comment (rebelchris) [inline src/routes/boot.ts:377]: @idoshamun just one mention for this experiment that it won't show to x% of selected numbers as we further down it here.
Just wanted to mention it for alignment.

File: src/routes/boot.ts [hunk 3]
```diff
@@ -346,6 +350,9 @@ const getAnonymousFirstVisit = async (trackingId: string) => {
   return finalValue;
 };
 
+// We released the firstVisit at July 10, 2023 - so to have a buffer time, the requirement is the day after.
+export const onboardingV2Requirement = new Date(2023, 6, 11);
```
Comment (idoshamun) [inline src/routes/boot.ts:354]: We will have to update it once we want to start the experiment. Just noting it here to remember

File: src/featureFlags.ts [hunk 4]
```diff
@@ -84,6 +85,27 @@ const getSubmitArticleState = (flags: IFlags, user: User) => {
   return flags?.submit_article;
 };
 
+interface UpdateParam {
+  feature: string;
+  replacement: Flag['value'];
+  onCheckValidity(value: Flag['value']): boolean;
```
Comment (idoshamun) [inline src/featureFlags.ts:91]: The name here is a bit confusing. I'd suggest changing it to `isUserApplicable`. It should return `true` if the we should keep the flag as is, and `false` if not. So in the case of the onboarding, if the user visited before the date they shouldn't be applicable and we return false.

File: src/featureFlags.ts [hunk 5]
```diff
@@ -84,6 +85,27 @@ const getSubmitArticleState = (flags: IFlags, user: User) => {
...
+
+export const adjustAnonymousFlags = (
+  flags: IFlags,
+  updates: UpdateParam[],
+): IFlags => {
+  const updatedFlags = { ...flags };
+
+  updates.forEach(({ feature, onCheckValidity, replacement }) => {
+    if (flags?.[feature]?.enabled && onCheckValidity(flags[feature].value)) {
+      updatedFlags[feature] = { enabled: true, value: replacement };
```
Comment (idoshamun) [inline src/featureFlags.ts:102]: If the user is not applicable for the flag, I suggest setting it to `enabled: false` at this point without value. This way we know in the data that this user is not part of an experiment

File: src/featureFlags.ts (hunk 4 above)
Comment (sshanzel) [reply src/featureFlags.ts:91]: Agreed, turning the value to false will be much better.

File: src/featureFlags.ts [hunk 6]
```diff
@@ -84,6 +85,28 @@ const getSubmitArticleState = (flags: IFlags, user: User) => {
...
+  flags: IFlags,
+  updates: UpdateParam[],
+): IFlags => {
+  const updatedFlags = { ...flags };
+
+  updates.forEach(({ feature, checkIsApplicable }) => {
+    if (!flags?.[feature]?.enabled) return;
+
+    if (!checkIsApplicable(flags[feature].value)) {
+      updatedFlags[feature] = { enabled: false, value: '' };
```
Comment (idoshamun) [inline src/featureFlags.ts:103]:
```suggestion
      updatedFlags[feature] = { enabled: false, value: null };
```

File: src/routes/boot.ts (hunk 3 above)
Comment (capJavert) [reply src/routes/boot.ts:354]: Could we not created flagsmith flag for these kind of things? Feels clunky to leave it like this.

**Review by rebelchris — APPROVED** [review]
Nothing from my side, just reminder to change the date when you decide to launch 🙏

**Comment by sshanzel** [comment]
Looking nice!

image

**Pull request merged**